### PR TITLE
[torchlib] Fix index_put: handle None cases

### PR DESCRIPTION
**Pull request by AyoubMDL**
This PR introduces support for `None` indices in the `index_put` function. If an index is None, it acts as a full slice (`:`).
### index_put Logic:
1. Construct index grid that contains all the indices to be updated
2. Reshapes the update values to match the computed indices.

### CODE REVIEW

**Comment by AyoubMDL** [comment]
@justinchuby PR is ready for review

**Comment by justinchuby** [comment]
Thank you! The failling test is unrelated

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 1]
```diff
@@ -4259,15 +4260,51 @@ def aten_index_put(
...
     """
+    # Pad indices with None so It has the same rank as self
+    self_rank = len(self.shape)
+    if len(indices) < self_rank:
+        indices = list(indices) + [None] * (self_rank - len(indices))
+
+    values_shape = values.shape.numpy()
+    # Pad values_shape with 1 so It has the same rank as self
+    if len(values_shape) < self_rank:
+        values_shape = (1,) * (self_rank - len(values_shape)) + values_shape
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:4271]:
```suggestion
        values_shape = [*((1,) * (self_rank - len(values_shape))), *values_shape]
```

File: tests/function_libs/torch_lib/extra_opinfo.py [hunk 2]
```diff
@@ -790,20 +790,27 @@ def sample_inputs_index_put(op_info, device, dtype, requires_grad, **kwargs):
...
+    make_arg = functools.partial(
+        torch_testing.make_tensor, device=device, dtype=dtype, requires_grad=requires_grad
     )
-    yield opinfo_core.SampleInput(data, indices, values)
+
+    data = make_arg((10, 3, 4))
+
+    cases = [
+        # Case 1: Full slices in dims 0 and 2, tensor index in dim 1
+        ([None, torch.arange(2, device=device), None], (10, 2, 4)),
```
[the remark this answers is not in the document]
Comment (AyoubMDL) [reply tests/function_libs/torch_lib/extra_opinfo.py:801]: I actually wanted to test that but It does behave differently, Here's an example:
If I do torch.arange to specify the indices it will behave like this:
```python
import torch
data = torch.randn((1, 8, 4))
updates = torch.randn((1, 2, 3))
data[:, [0, 1], [1, 2, 3]] = updates
```
This results in an error: `shape mismatch: indexing tensors could not be broadcast together with shapes [2], [3]`

However, the following approach works:
```python
import torch
data = torch.randn((1, 8, 4))
updates = torch.randn((1, 2, 3))
data[:, 0:2, 1:4] = updates
```
The issue is that I don't know how to specify indices in the unit test in this way.

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 3]
```diff
@@ -1647,7 +1647,8 @@ def aten_chunk(self: TTensor, chunks: int, dim: int = 0) -> Sequence[TTensor]:
     dim_size = op.Gather(self_shape, dim, axis=0)
     # Compute size/chunk to get the number of data in one chunk
     num_per_chunk = op.Div(dim_size, chunks)
-    num_per_chunk = op.Cast(op.Mod(dim_size, chunks) > 0, to=INT64.dtype) + num_per_chunk  # type: ignore[operator]
+    num_per_chunk = op.Cast(op.Mod(dim_size, chunks) > 0, to=INT64.dtype) + \
```
[the remark this answers is not in the document]
Comment (AyoubMDL) [reply onnxscript/function_libs/torch_lib/ops/core.py:1650]: oops (it was the linter)

File: tests/function_libs/torch_lib/extra_opinfo.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AyoubMDL) [reply tests/function_libs/torch_lib/extra_opinfo.py]: Added the cases, tell me if other cases are needed

**Comment by AyoubMDL** [comment]
Force pushed as I changed a little bit the logic. Also added more test cases.

**Comment by justinchuby** [comment]
Could you take a look at the falling CI, and we can merge this PR? Thanks!

**Comment by AyoubMDL** [comment]
> Could you take a look at the falling CI, and we can merge this PR? Thanks!

it was missing a case with multidimensional index (added a test for that case)

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 4]
```diff
@@ -4303,17 +4305,21 @@ def _make_reshape_list_broadcastable(reshape_list, values_shape):
         while len(reshape_list) > len(values_shape) and 1 in reshape_list:
             reshape_list.remove(1)
 
+        # Or add ones until the rank of reshape_list matches values_shape.
+        while len(reshape_list) < len(values_shape):
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:4309]: Possible to compute the difference and extend directly to avoid a while loop?

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 5]
```diff
@@ -4298,14 +4300,82 @@ def aten_index_put(
...
+
+    index_vectors = []
+    for i in range(self_rank):
+        if indices[i] is None:
+            # For a full slice along dim i, create a range index [0, self.shape[i]).
+            idx = op.Range(start=0, limit=self.shape[i], delta=1)
+            reshape_update = self.shape[i]
+        else:
+            idx = indices[i]
+            reshape_update = np.prod(idx.shape).item()
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:4342]: We can use math.prod instead
```suggestion
            reshape_update = math.prod(idx.shape)
```

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 6]
```diff
@@ -4298,14 +4300,82 @@ def aten_index_put(
...
+        indices = list(indices) + [None] * (self_rank - len(indices))
+
+    # Get values shape
+    values_shape = tuple(values.shape)
+
+    index_vectors = []
+    for i in range(self_rank):
+        if indices[i] is None:
+            # For a full slice along dim i, create a range index [0, self.shape[i]).
+            idx = op.Range(start=0, limit=self.shape[i], delta=1)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:4338]:
```suggestion
            idx = op.Range(0, op.Shape(self, i, i+1), 1)
```

Do we want this to be dynamic? If so

Comment (AyoubMDL) [reply onnxscript/function_libs/torch_lib/ops/core.py:4338]: Sure

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 4 above)
Comment (AyoubMDL) [reply onnxscript/function_libs/torch_lib/ops/core.py:4309]: Actually, I don't need that part as I shrink the values_shape before (so always len(reshape_list) >= len(values_shape))

File: onnxscript/function_libs/torch_lib/ops/core.py [hunk 7]
```diff
@@ -4298,14 +4298,78 @@ def aten_index_put(
     <https://github.com/pytorch/pytorch/blob/main/torch/onnx/symbolic_opset11.py#L212>`_.
     """
 
-    # TODO(justinchuby): Handle when indicies has more than one element
-    index = indices[0]
-    new_index = op.Unsqueeze(index, [-1])
+    def _make_reshape_list_broadcastable(reshape_list, values_shape):
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_lib/ops/core.py:4301]: Could you annotate the type of this function? I am a little lost on which one is expected contain dynamic values and which one doesn't.

In general, if x is a dynamic value (tensor), we can only take the rank of it by `len(x.shape)` but not individual shapes via `x.shape` (due to implementation constraints). To keep the dim dynamic we need `op.Shape(x, start=, end=)`. I see a few places where `shape` is assume to contain static (int) dimensions. Could you fix those places? Thanks!

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 6 above)
Comment (AyoubMDL) [reply onnxscript/function_libs/torch_lib/ops/core.py]: Not sure why the test fails [here](https://github.com/microsoft/onnxscript/actions/runs/13590681526/job/37998239141?pr=2061) and passes on other setups. I think because Shape returns a tensor and not a scalar which what Range expect. Maybe adding a Gather would solve that.

**Comment by justinchuby** [comment]
For range, I think we need to Squeeze the tensor to remove its dimension.

File: onnxscript/function_libs/torch_lib/ops/core.py (hunk 7 above)
Comment (AyoubMDL) [reply onnxscript/function_libs/torch_lib/ops/core.py:4301]: Should I rewrite the logic of this function assuming that values are dynamic ?

Comment (AyoubMDL) [reply onnxscript/function_libs/torch_lib/ops/core.py:4301]: Ok, I'll convert this PR to draft while working on it

**Comment by AyoubMDL** [comment]
@justinchuby I wasn’t able to make this work on symbolic tensors. Please feel free to close the draft or continue from here if helpful.

**Comment by justinchuby** [comment]
Thanks. Could you share the issues you have seen? I can take it from here

**Review by justinchuby — APPROVED** [review]

**Comment by justinchuby** [comment]
Will merge as-is and create follow up improvements

**Comment by AyoubMDL** [comment]
> Thanks. Could you share the issues you have seen? I can take it from here

- I didn't know how to do the part where I construct the reshape_list using only onnx operations.
- I also tried to adapt the implementation from pytorch https://github.com/pytorch/pytorch/blob/6c3492b49118d9407db39fc5b134c0f96739b831/torch/onnx/symbolic_opset11.py#L206 but there logic doesn't work when I want to handle None indices.
